src/Api/Auth/Controllers/WebAuthnController.cs

```diff
@@ -14,67 +17,101 @@ namespace Bit.Api.Auth.Controllers;
         }
-        return new ListResponseModel<TwoFactorWebAuthnResponseModel>(new List<TwoFactorWebAuthnResponseModel> { });
+
+        var credentials = await _credentialRepository.GetManyByUserIdAsync(user.Id);
+
+        return new ListResponseModel<WebAuthnCredentialResponseModel>(credentials.Select(c => new WebAuthnCredentialResponseModel(c)));
     }
 
     [HttpPost("options")]
     [ApiExplorerSettings(IgnoreApi = true)] // Disable Swagger due to CredentialCreateOptions not converting properly
-    public async Task<CredentialCreateOptions> PostOptions()
+    public async Task<WebAuthnCredentialCreateOptionsResponseModel> PostOptions([FromBody] SecretVerificationRequestModel model)
     {
-        var user = await _userService.GetUserByPrincipalAsync(User);
-        if (user == null)
-        {
-            throw new UnauthorizedAccessException();
-        }
+        var user = await CheckAsync(model);
+        var options = await _userService.StartWebAuthnLoginRegistrationAsync(user);
+
+        var tokenable = new WebAuthnCredentialCreateOptionsTokenable(user, options);
+        var token = _createOptionsDataProtector.Protect(tokenable);
 
-        var reg = await _userService.StartWebAuthnLoginRegistrationAsync(user);
-        return reg;
+        return new WebAuthnCredentialCreateOptionsResponseModel
+        {
+            Options = options,
+            Token = token,
+        };
     }
 
     [HttpPost("")]
-    // TODO: Create proper models for this call
-    public async Task<TwoFactorWebAuthnResponseModel> Post([FromBody] TwoFactorWebAuthnRequestModel model)
+    public async Task Post([FromBody] WebAuthnCredentialRequestModel model)
     {
         var user = await _userService.GetUserByPrincipalAsync(User);
-        if (user == null)
+
+        var tokenable = _createOptionsDataProtector.Unprotect(model.Token);
+        if (!tokenable.TokenIsValid(user))
         {
-            throw new UnauthorizedAccessException();
+            throw new BadRequestException("The token associated with your request is expired. A valid token is required to continue.");
         }
 
-        var success = await _userService.CompleteWebAuthLoginRegistrationAsync(user, model.Name, model.DeviceResponse);
+        var success = await _userService.CompleteWebAuthLoginRegistrationAsync(user, model.Name, tokenable.Options, model.DeviceResponse);
         if (!success)
         {
             throw new BadRequestException("Unable to complete WebAuthn registration.");
         }
-        var response = new TwoFactorWebAuthnResponseModel(user);
-        return response;
     }
 
-    [HttpDelete("{id}")]
     [HttpPost("{id}/delete")]
-    public async Task Delete(string id)
+    public async Task Delete(string id, [FromBody] SecretVerificationRequestModel model)
+    {
+        var user = await CheckAsync(model);
+        var credential = await _credentialRepository.GetByIdAsync(new Guid(id), user.Id);
```

Comment:
❓ Can we just accept this as a `Guid` in the method signature?

Reply:
Fixed!